packages/assets-controllers/src/MultichainAssetsController/MultichainAssetsController.ts

```diff
@@ -0,0 +1,395 @@
+    }
+
+    // Get assets list
+    if (account.metadata.snap) {
+      const assets = await this.#getAssets(
+        account.id,
+        account.metadata.snap.id,
+      );
+      const assetsWithoutMetadata = assets.filter(
+        (asset) => !this.state.metadata[asset],
+      );
+      const snaps = this.#getAllSnaps();
+
+      const permissions = snaps.map((snap) =>
+        this.#getSnapsPermissions(snap.id),
+      );
+
+      // Mock start To be removed once the above is implemented
+      permissions.forEach((singlePermission) => {
+        (singlePermission as unknown as AssetEndowment) = {
+          ...singlePermission,
+          'endowment:assets': {
+            scopes: ['bip122:000000000019d6689c085ae165831e93'],
+          },
+        };
+      });
+      (permissions[0] as unknown as AssetEndowment) = {
+        ...permissions[0],
+        'endowment:assets': {
+          scopes: ['solana:EtWTRABZaYq6iMfeYKouRu166VU2xqa1'],
+        },
+      };
+      // Mock End To be removed once the above is implemented
+
+      // Identify the correct snap that has the right endowment:assets permission
+      // TODO: create a mapping of the assets to the snapId based on the permission['endowment:assets']?.scopes from permissions array and the assets[0].split('/')[0] from assets array
+      const mapAssetsToSnapId = new Map<CaipAssetType, string[]>();
+      assets.forEach((asset) => {
+        const snapIds: string[] = [];
+        permissions.forEach((permission: AssetEndowment, index: number) => {
+          if (
+            permission['endowment:assets']?.scopes.includes(asset.split('/')[0])
+          ) {
+            snapIds.push(snaps[index].id);
+          }
+        });
+        mapAssetsToSnapId.set(asset, snapIds);
+      });
+      // should take the first snapId from the mapAssetsToSnapId and use it to get the assets
+
+      // call the snap to get the metadata
+      if (assetsWithoutMetadata.length > 0) {
+        const metadata = await this.#getMetadata(assetsWithoutMetadata);
+
+        const newMetadata = {
+          ...this.state.metadata,
+          ...metadata.assets,
+        };
+        this.update((state) => {
+          state.metadata = newMetadata;
+        });
+      }
+      this.update((state) => {
+        state.allNonEvmTokens[account.id] = assets;
+      });
+    }
+  }
+
+  /**
+   * Handles changes when a new account has been removed.
+   *
+   * @param accountId - The new account id being removed.
+   */
+  async #handleOnAccountRemoved(accountId: string): Promise<void> {
+    const selectedAccounts = this.messagingSystem.call(
+      'AccountsController:listMultichainAccounts',
+    );
+
+    const nonEvmAccounts = selectedAccounts.filter((account) =>
+      this.#isNonEvmAccount(account),
+    );
+    const account: InternalAccount | undefined = nonEvmAccounts.find(
+      (multichainAccount) => multichainAccount.id === accountId,
+    );
+    if (!account) {
+      return;
+    }
```

Comment:
I wonder if that's really necessary?

We could check if `accountId` exists in `state.allNonEvmTokens`, no?

Reply:
Yes i think that should be sufficient

Reply:
updated the logic here [9bbe320](https://github.com/MetaMask/core/pull/5138/commits/9bbe320a21e9abea4db529ccdeb7cd413e2c2d3d)